### DISCUSSION TITLE
PSK lookup issue

### DISCUSSION THREAD

**Discussion by mohanraj-cit** [strongswan/strongswan#669] [Q&A] [answered]
Hello there,

I am new to the mailing list and I have decent understanding of Strongswan project.

I am having a strange PSK lookup issue when I have 2 tunnels going from 2 different interfaces (side A)  to same interface (side B)  and I would like to know if it is a ID/PSK lookup bug or any configuration issue. 

Setup 
=====
-	Side A , 1 Linux running Strongsawn 5.8.4 with 2 interfaces (wan0 and wan1) 
-	Side B , 1 Linux running Strongswan 5.9.0 with 1 interface (eth1) 
-	“wan0_to_linux “ tunnel (ikev2) from Side A wan0 -> Side B eth1
-	“wan1_to_linux“ tunnel (Ikev2) from Side A wan1 to Side B eth1
-	Unique PSKs are configured for both tunnels

Problem
======
-	Both tunnels are established successfully when I establish them one after the other (logs below)
-	But wan0_to_linux tunnels never comes UP when I establish them together (logs below)
o	When I look at the DBG4 logs, it looks like a wrong PSK is looked up and AUTH fails (logs below)

Config
=====

Side A
====

cat /etc/swanctl/conf.d/swanctl.conf 
connections {
        wan0_to_linux { 
                local_addrs = 10.80.139.194
                remote_addrs = 10.80.139.183
                version = 2
                aggressive = yes
                proposals = aes256-sha1-modp1024, aes128-sha1-modp1024
                rekey_time = 360m
                over_time = 1m
                dpd_delay = 10s
                dpd_timeout = 30s
                dscp = 0
                mobike = no
                local {
                        auth = psk
                        id = 10.80.139.194
                }
                remote {
                        auth = psk
                        id = 10.80.139.183
                }
                children {
                        child-a {
                                local_ts = 0.0.0.0/0, ::/0
                                remote_ts = 0.0.0.0/0, ::/0
                                esp_proposals = aes256-sha1-modp1024, aes128-sha1-modp1024
                                life_time = 360m
                                rekey_time = 0 
                                rand_time = 0
                                mode = tunnel
                                dpd_action = clear
                                policies = no
                        }
                }
        }
        wan1_to_linux { 
                local_addrs = 10.80.141.51
                remote_addrs = 10.80.139.183
                version = 2
                aggressive = yes
                proposals = aes256-sha1-modp1024, aes128-sha1-modp1024
                rekey_time = 360m
                over_time = 1m
                dpd_delay = 10s
                dpd_timeout = 30s
                dscp = 0
                mobike = no
                local {
                        auth = psk
                        id = 10.80.141.51
                }
                remote {
                        auth = psk
                        id = 10.80.139.183
                }
                children {
                        child-a {
                                local_ts = 0.0.0.0/0, ::/0
                                remote_ts = 0.0.0.0/0, ::/0
                                esp_proposals = aes256-sha1-modp1024, aes128-sha1-modp1024
                                life_time = 360m
                                rekey_time = 0 
                                rand_time = 0
                                mode = tunnel
                                dpd_action = clear
                                policies = no
                        }
                }
        }
}
secrets {
        ike-wan0_to_linux {
                id = 10.80.139.183
                secret = wan0_to_linux
        }
        ike-wan1_to_linux {
                id = 10.80.139.183
                secret = wan1_to_linux
        }
}

Side B
====

cat /usr/local/etc/ipsec.conf
config setup

conn %default
        ikelifetime=1440m
        lifetime=1440m
        keyingtries=1
        authby=secret
        keyexchange=ikev2
        mobike=no
        dpddelay=20
        dpdtimeout=100
        dpdaction=clear

conn net-net-139
        left=10.80.139.183
        leftid=10.80.139.183
        leftsubnet=10.23.25.0/24
        rightsubnet=10.23.60.0/24
        auto=add
        ike=aes256-sha1-modp1024, aes128-sha1-modp1024
        esp=aes256-sha1-modp1024, aes128-sha1-modp1024

cat /usr/local/etc/ipsec.secrets
10.80.139.194 : PSK "wan0_to_linux"
10.80.141.51  : PSK "wan1_to_linux"

When I try to establish tunnels one after the other, tunnel s come up fine as expected.
=======================================================

swanctl -t --ike wan0_to_linux; swanctl -i --ike wan0_to_linux --child child-a
[IKE] deleting IKE_SA wan0_to_linux[921] between 10.80.139.194[10.80.139.194]...10.80.139.183[10.80.139.183]
[IKE] sending DELETE for IKE_SA wan0_to_linux[921]
[ENC] generating INFORMATIONAL request 2 [ D ]
[NET] sending packet: from 10.80.139.194[500] to 10.80.139.183[500] (76 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.139.194[500] (76 bytes)
[ENC] parsed INFORMATIONAL response 2 [ ]
[IKE] IKE_SA deleted
terminate completed successfully
[IKE] initiating IKE_SA wan0_to_linux[922] to 10.80.139.183
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 10.80.139.194[500] to 10.80.139.183[500] (380 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.139.194[500] (344 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
[IKE] authentication of '10.80.139.194' (myself) - '10.80.139.183' (remote) with pre-shared key
[IKE] establishing CHILD_SA child-a{931}
[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
[NET] sending packet: from 10.80.139.194[500] to 10.80.139.183[500] (364 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.139.194[500] (220 bytes)
[ENC] parsed IKE_AUTH response 1 [ IDr AUTH SA TSi TSr N(AUTH_LFT) ]
[IKE] authentication of '10.80.139.183' with pre-shared key successful
[IKE] IKE_SA wan0_to_linux[922] established between 10.80.139.194[10.80.139.194]...10.80.139.183[10.80.139.183]
[IKE] scheduling rekeying in 21598s
[IKE] maximum IKE_SA lifetime 21658s
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
[IKE] CHILD_SA child-a{931} established with SPIs c3a74841_i c81358eb_o and TS 10.23.60.0/24 === 10.23.25.0/24
[IKE] received AUTH_LIFETIME of 85832s, scheduling reauthentication in 85772s
initiate completed successfully

swanctl -t --ike wan1_to_linux; swanctl -i --ike wan1_to_linux --child child-a
[IKE] deleting IKE_SA wan1_to_linux[924] between 10.80.141.51[10.80.141.51]...10.80.139.183[10.80.139.183]
[IKE] sending DELETE for IKE_SA wan1_to_linux[924]
[ENC] generating INFORMATIONAL request 2 [ D ]
[NET] sending packet: from 10.80.141.51[500] to 10.80.139.183[500] (76 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.141.51[500] (76 bytes)
[ENC] parsed INFORMATIONAL response 2 [ ]
[IKE] IKE_SA deleted
terminate completed successfully
[IKE] initiating IKE_SA wan1_to_linux[925] to 10.80.139.183
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 10.80.141.51[500] to 10.80.139.183[500] (380 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.141.51[500] (344 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
[IKE] authentication of '10.80.141.51' (myself) - '10.80.139.183' (remote) with pre-shared key
[IKE] establishing CHILD_SA child-a{934}
[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
[NET] sending packet: from 10.80.141.51[500] to 10.80.139.183[500] (364 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.141.51[500] (220 bytes)
[ENC] parsed IKE_AUTH response 1 [ IDr AUTH SA TSi TSr N(AUTH_LFT) ]
[IKE] authentication of '10.80.139.183' with pre-shared key successful
[IKE] IKE_SA wan1_to_linux[925] established between 10.80.141.51[10.80.141.51]...10.80.139.183[10.80.139.183]
[IKE] scheduling rekeying in 21598s
[IKE] maximum IKE_SA lifetime 21658s
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
[IKE] CHILD_SA child-a{934} established with SPIs ce63fd4c_i cb8145d0_o and TS 10.23.60.0/24 === 10.23.25.0/24
[IKE] received AUTH_LIFETIME of 85722s, scheduling reauthentication in 85662s
initiate completed successfully

When I try to establish the tunnels together, wan0_to_linux never comes up.
=================================================

swanctl -i --ike wan0_to_linux --child child-a ;  swanctl -i wan1_to_linux --child child-a
[IKE] initiating IKE_SA wan0_to_linux[912] to 10.80.139.183
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 10.80.139.194[500] to 10.80.139.183[500] (380 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.139.194[500] (344 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
[IKE] authentication of '10.80.139.194' (myself) - '10.80.139.183' (remote) with pre-shared key
[IKE] establishing CHILD_SA child-a{920}
[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
[NET] sending packet: from 10.80.139.194[500] to 10.80.139.183[500] (364 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.139.194[500] (76 bytes)
[ENC] parsed IKE_AUTH response 1 [ N(AUTH_FAILED) ]
[IKE] received AUTHENTICATION_FAILED notify error
initiate failed: establishing CHILD_SA 'child-a' failed
[IKE] initiating IKE_SA wan1_to_linux[913] to 10.80.139.183
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 10.80.141.51[500] to 10.80.139.183[500] (380 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.141.51[500] (344 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
[IKE] authentication of '10.80.141.51' (myself) - '10.80.139.183' (remote) with pre-shared key
[IKE] establishing CHILD_SA child-a{921}
[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
[NET] sending packet: from 10.80.141.51[500] to 10.80.139.183[500] (364 bytes)
[NET] received packet: from 10.80.139.183[500] to 10.80.141.51[500] (220 bytes)
[ENC] parsed IKE_AUTH response 1 [ IDr AUTH SA TSi TSr N(AUTH_LFT) ]
[IKE] authentication of '10.80.139.183' with pre-shared key successful
[IKE] IKE_SA wan1_to_linux[913] established between 10.80.141.51[10.80.141.51]...10.80.139.183[10.80.139.183]
[IKE] scheduling rekeying in 21559s
[IKE] maximum IKE_SA lifetime 21619s
[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA1_96/NO_EXT_SEQ
[IKE] CHILD_SA child-a{921} established with SPIs c8b199eb_i c264bfb2_o and TS 10.23.60.0/24 === 10.23.25.0/24
[IKE] received AUTH_LIFETIME of 85797s, scheduling reauthentication in 85737s
initiate completed successfully
[admin@dev03-ecvf root]#

Logs
====
Looking at the DBG4 logs, for wan1_to_linux tunnel,  “wan1_to_linux” PSK is looked up as expected and tunnel comes UP.

Oct 14 23:07:22 2021 dev03-ecvf IKE: 15[IKE] <wan1_to_linux|936> secret => 13 bytes @ 0x7fabb8003f80
Oct 14 23:07:22 2021 dev03-ecvf IKE: 15[IKE] <wan1_to_linux|936>    0: 77 61 6E 31 5F 74 6F 5F 6C 69 6E 75 78           wan1_to_linux

**But for wan0_to_linux tunnel, a wrong PSK is looked up and AUTH fails**

Oct 14 23:07:22 2021 IKE: 06[IKE] <**wan0_to_linux**|935> secret => 13 bytes @ 0x7fabb8003f80
Oct 14 23:07:22 2021 IKE: 06[IKE] <**wan0_to_linux**|935>    0: 77 61 6E 31 5F 74 6F 5F 6C 69 6E 75 78           **wan1_to_linux**

Questions
=======
-	Looking at the source code (vici_cred.c, cred_manager.c, identification.c, psk_authenticator.c), it is not clear how exactly PSK lookup is performed. Is there a document which explains PSK/ID lookup ?
-	Is there a way I can dump the PSKs and associated IDs ?
-	Could you please let me know if my secret block is configured correctly (especially the ID ownership for PSK) ?

Any help is much appreciated!

Thanks in advance.
Mohan

**Comment by tobiasbrunner** [comment, marked as answer]
>   it is not clear how exactly PSK lookup is performed. Is there a document which explains PSK/ID lookup ?

It's quite simple, shared secrets have to match identities (local and remote IKE identities in case of PSKs). If multiple secrets match the local and remote identities equally well (i.e. match the same number in the same way e.g. without wildcards), the order is undefined (depends on the credential sets that provide them, could change whenever e.g. configs are reloaded). As initiator, only the best/first match will be used. As responder, all matching PSKs will be tried.

>  Is there a way I can dump the PSKs and associated IDs ?

At what point?

> Could you please let me know if my secret block is configured correctly (especially the ID ownership for PSK) ?

If you want to use different secrets for different connections, configuring the same identity for both PSKs clearly makes no sense as initiator. So either add the local identity, which is different, too, or only configure that instead of the shared remote identity. Or make the remote identity different too (doesn't have to be an IP address).

But what's the point of using different PSKs for two connections between the same peers in the first place?

**Reply by mohanraj-cit** [reply]
Thank you for your prompt response Tobias.

I still don't understand this lookup logic. I have some follow up questions.

>It's quite simple, shared secrets have to match identities (local and remote IKE identities in case of PSKs).

If I look at the simple swanctl PSK examples (carol), https://www.strongswan.org/testing/testresults/ikev2/rw-psk-ipv4/carol.swanctl.conf 

```
secrets {
   ike-moon {
      id = 192.168.0.1 
      secret = 0sFpZAZqEN6Ti9sqt4ZP5EWcqx
   }
}
```
PS: secret 0sFpZAZqEN6Ti9sqt4ZP5EWcqx is associated only with ID 192.168.0.1 (moon) and NOT with ID 192.168.0.100 (carol)
When carol initiates, how does PSK lookup happen when there is only 1 ID specified in secrets block
- Will this ID (in secret block) be used as source ID or destination ID ?
- Will 2 lookup happen like (any, id) and (id, any) ?

>If you want to use different secrets for different connections, configuring the same identity for both PSKs clearly makes no sense as initiator. So either add the local identity, which is different, too, or only configure that instead of the shared remote identity. Or make the remote identity different too (doesn't have to be an IP address).

yes. I want to use different PSKs for each tunnel. Are you suggesting, I use both IDs in the secret block like https://www.strongswan.org/testing/testresults/ikev2/net2net-psk/moon.swanctl.conf ?

```
secrets {
    ike-wan0_to_linux {
        id-local = 10.80.141.51
        id-remote = 10.80.139.183
        secret = wan0_to_linux
    }
}
```
 When 2 IDs are specified for 1 PSK in secrets block, when the node initiates, how does PSK lookup happen ?
- Will the IDs be used as source ID and destination ID (id-local, id-remote) ?

>But what's the point of using different PSKs for two connections between the same peers in the first place?

Some cloud security vendors have restriction to use unique PSK for every tunnel/connection.

**Reply by tobiasbrunner** [reply]
>  When carol initiates, how does PSK lookup happen when there is only 1 ID specified in secrets block

If you look at the `remote` section, you see that `id = 192.168.0.1` is configured.

> When 2 IDs are specified for 1 PSK in secrets block, when the node initiates, how does PSK lookup happen ?

As I already described. The local and remote IKE identities of the connection are matched against the identities configured for the shared secret.

**Reply by mohanraj-cit** [reply]
Thank you for your time and answers Tobias.
